### ISSUE TITLE
Loss function nan

### ISSUE THREAD

**Issue by Nicolas99-9** [keras-team/keras#8508]
I tried to create a simple neural network but the loss function is always nan.
My data is a matrix with the shape (84906, 23) 
The labels can have two values (1 or 2).
My code

`import numpy as np

def f1_score(y_true, y_pred):
    # Count positive samples.
    c1 = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
    c2 = K.sum(K.round(K.clip(y_pred, 0, 1)))
    c3 = K.sum(K.round(K.clip(y_true, 0, 1)))

    # If there are no true samples, fix the F1 score at 0.
    if c3 == 0:
        return 0

    # How many selected items are relevant?
    precision = c1 / c2

    # How many relevant items are selected?
    recall = c1 / c3

    # Calculate f1_score
    f1_score = 2 * (precision * recall) / (precision + recall)
    return f1_score 

print(X_train.shape)
y_train2 =  to_categorical(y_train-1)
model = Sequential()
model.add(Dense(50, input_dim=23,activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(30,activation='softmax'))
model.add(Dropout(0.5))
model.add(Dense(2,activation='relu'))
print("e4")
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy',f1_score])
print("e5")
model.fit(X_train, y_train2, batch_size=128,  verbose=2)
import numpy as np

def f1_score(y_true, y_pred):
    # Count positive samples.
    c1 = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
    c2 = K.sum(K.round(K.clip(y_pred, 0, 1)))
    c3 = K.sum(K.round(K.clip(y_true, 0, 1)))

    # If there are no true samples, fix the F1 score at 0.
    if c3 == 0:
        return 0

    # How many selected items are relevant?
    precision = c1 / c2

    # How many relevant items are selected?
    recall = c1 / c3

    # Calculate f1_score
    f1_score = 2 * (precision * recall) / (precision + recall)
    return f1_score 

print(X_train.shape)
y_train2 =  to_categorical(y_train-1)
model = Sequential()
model.add(Dense(50, input_dim=23,activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(30,activation='softmax'))
model.add(Dropout(0.5))
model.add(Dense(2,activation='relu'))
print("e4")
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy',f1_score])
print("e5")
model.fit(X_train, y_train2, batch_size=128,  verbose=2)`

The output:

`Epoch 1/10
1s - loss: nan - acc: 0.9682 - f1_score: nan
Epoch 2/10
1s - loss: nan - acc: 0.9689 - f1_score: nan
Epoch 3/10
1s - loss: nan - acc: 0.9689 - f1_score: nan
Epoch 4/10
1s - loss: nan - acc: 0.9689 - f1_score: nan
Epoch 5/10
1s - loss: nan - acc: 0.9689 - f1_score: nan
Epoch 6/10
1s - loss: nan - acc: 0.9689 - f1_score: nan
Epoch 7/10
1s - loss: nan - acc: 0.9689 - f1_score: nan
Epoch 8/10
1s - loss: nan - acc: 0.9689 - f1_score: nan
Epoch 9/10
1s - loss: nan - acc: 0.9689 - f1_score: nan
Epoch 10/10
1s - loss: nan - acc: 0.9689 - f1_score: nan`

**Comment by DomHudson** [comment]
Hi,
as your data has either class 1 or class 2 it's a binary problem. Therefore you should change your targets so to 1 and 0 and use this as your final layer:
```
Dense(1, activation = 'sigmoid')
```
You should also use 'binary_crossentropy'.
